R/pipeline.R

```diff
@@ -0,0 +1,221 @@
+#' task-specific directory (`model.rds`).
+#' - **Samples**: Parquet files containing the model's sample outputs are saved
+#' in a `samples` subdirectory, named using the `task_id` (e.g.,
+#' `task_id.parquet`).
+#' - **Summaries**: Parquet files summarizing the model's results are saved in
+#' a `summaries` subdirectory, also named using the `task_id` (e.g.,
+#' `task_id.parquet`).
+#' - **Logs**: A `logs.txt` file is generated in the task directory, capturing
+#' both console and error messages.
+#'
+#' The output directory structure will follow this format:
+#' ```
+#' <output_dir>/
+#' └── <job_id>/
+#'     ├── samples/
+#'     │   └── <task_id>.parquet
+#'     ├── summaries/
+#'     │   └── <task_id>.parquet
+#'     └── tasks/
+#'         └── <task_id>/
+#'             ├── model.rds
+#'             └── logs.txt
+#' ```
+#'
+#' @return The function does not return a value directly. However, the output
+#' directory will contain the following files:
+#' - Model RDS file (`model.rds`)
+#' - Sample output in Parquet format (`<task_id>.parquet` in the `samples/`
+#' directory)
+#' - Summary output in Parquet format (`<task_id>.parquet` in the `summaries/`
+#' directory)
+#' - Log file (`logs.txt`) in the task directory
+#'
+#' @rdname pipeline
+#' @export
+run_pipeline <- function(config_path,
+                         blob_storage_container = NULL,
+                         output_dir = "/") {
+  # TODO: Add config reader here
+  config <- jsonlite::read_json(config_path,
+    simplifyVector = TRUE
+  )
+
+  write_output_dir_structure(
+    output_dir = output_dir,
+    job_id = config[["job_id"]],
+    task_id = config[["task_id"]]
+  )
+
+  # Set up logs
+  logfile_path <- file.path(
+    output_dir,
+    config[["job_id"]],
+    "tasks",
+    config[["task_id"]]
+  )
+  logfile_connection <- file(file.path(logfile_path, "logs.txt"), open = "wt")
+  sink(
+    logfile_connection,
+    type = "output",
+    append = TRUE,
+    # Send output to logs and to console
+    split = TRUE
+  )
+  sink(
+    logfile_connection,
+    type = "message",
+    append = TRUE
+  )
+  on.exit(sink(file = NULL))
+  cli::cli_alert_info("Starting run at {Sys.time()}")
+  cli::cli_alert_info("Using job id {.field {config[['job_id']]}}")
+  cli::cli_alert_info("Using task id {.field {config[['task_id']]}}")
+
+  # Errors within `process_pipeline()` are downgraded to warnings so
+  # they can be logged and stored in Blob. If there is an error,
+  # `pipeline_success` is set to false, which will be stored in the
+  # metadata in the next PR.
+  pipeline_success <- rlang::try_fetch(
+    process_pipeline(config, output_dir),
+    error = function(con) {
+      cli::cli_warn("Pipeline run failed",
+        parent = con,
+        class = "Run_failed"
+      )
+      FALSE
+    }
+  )
+
+  # TODO: Move metadata to outer wrapper
+  cli::cli_alert_info("Finishing run at {Sys.time()}")
+  invisible(pipeline_success)
+}
+
+#' Run the Model Fitting Process
```

Comment:
If this is called `process_pipeline()` then I think it's inconsistent with the title being "Run the Model Fitting Process"

Reply:
I don't love the name either but couldn't think of something better. Do you have a suggestion that makes more sense to you?

Reply:
* `run_pipeline()` has title "Run an Rt Estimation Model Pipeline" and calls
* `process_pipeline()` has the title "Run the Model Fitting Process" and calls
* `fit_model()` as well as other things

I guess this naming question might be pointing as this modularisation of things being questionable.

What benefit comes from there being a `process_pipeline` function rather than it all being in `run_pipeline`?

I guess the prompt is to think about the ways in which we might change the code and how this structure sets us up for that.

In some ways I might be more up for describing "running the pipeline" as the `process_pipeline()` part and the `run_pipeline()` function is related to in particular running it on Blob.

I might try asking AI for some names.

Reply:
Something like `deploy_pipeline()` for the top level one seems reasonable to me. It could be S3.

Reply:
(And then `run_pipeline()` as the mid level one)

Reply:
Or `deploy_pipeline_blob()` as top level.

Reply:
I don't really agree with the above. I think there's some miscommunication about purpose here -- which I think comes back to the semantics around naming rather than structural changes.

> What benefit comes from there being a process_pipeline function rather than it all being in run_pipeline?

As I write in the starter comment

> I split the pipeline into two phases (1) In inner "process pipeline" (taking suggestions on the name) that handles the business logic and (2) an outer pipeline wrapper that handles interface with external services (logging, Azure, error handling). Errors in (1) can be handled and will appear in logs and metadata. Errors in (2) are catastrophic and will lead to unhandled failures.

The point being, this two-layer format isn't decorative; it's structural. It enables logging and error handling.  

> In some ways I might be more up for describing "running the pipeline" as the process_pipeline() part and the run_pipeline() function is related to in particular running it on Blob.

I don't agree with this comment either. This structure sets us up to run (1) totally locally (2) locally while reading/writing from blob or (3) in Batch reading/writing from blob. We don't have access to any other cloud environments and that's unlikely to change. Optimizing for generality in case we might want to plug into a different cloud provider someday strikes me as premature optimization.

But none of the above means that the current wrapper function names are good -- it probably indicates the opposite... 

I'll take another crack at them and update here.

Reply:
Could try writing down the combination of functions you'd call to do each of:

(1) totally locally (2) locally while reading/writing from blob or (3) in Batch reading/writing from blob.

Then use this to base the function names on. (This is where I was going with "`deploy_pipeline_blob`" type ideas.)

Reply:
Possible `outer_pipeline` and `inner_pipeline`.

Key distinction is "interface with external services" (outer) versus "this package" (inner).

Happens to be that that corresponds with "catastrophic errors" (outer) versus "logged errors" (inner).

Reply:
Ah ok.

Currently we can only do totally locally.

That's what I meant by

> I split the pipeline into two phases (1) In inner "process pipeline" (taking suggestions on the name) that handles the business logic and (2) an outer pipeline wrapper that handles interface with external services (logging, Azure, error handling). 
> ...
> I only handle phase (1) here. Phase (2) will be handled in another PR

I would really like to leave (2) out of scope here and focus on it in the next PR. Does that work for you? It should be open today or Monday.

Reply:
I think a benefit of "inner" and "outer" is that they are easy to talk to people about. If we are talking and you say let's run the pipeline or let's deploy the pipeline or let's process the pipeline it gets easy to be confused what you mean. If it's "[...] the inner pipeline" or "[...] the outer pipeline" it's clear to me what you mean.

Reply:
Ok another conversation but arguably something like this "inner pipeline" is what the CFA EpiNow2 pipeline should mean. We are going to have lots of `inner_pipelines` eventually in different repos. The outer pipeline part should be the same between them all? If so I think we should design with it in mind that this part of the package is eventually moved to some other package which interfaces with many model specific pipelines (like EpiNow2 pipeline).

Feel free to take or leave what I'm saying, just thinking out loud.

Reply:
> Optimizing for generality in case we might want to plug into a different cloud provider someday strikes me as premature optimization.

Along these lines then we should be optimising for generality incase we are adding another model pipeline (which we do plan to do AFAIK). So another way to say is that this `deploy_` word is pipeline agnostic and its own package.

Reply:
I agree that `run_pipeline` (inner function) and `deploy_azure_run()` or something along those lines could be better names but we can make this a fix later.

Reply:
Suggest making an issue to summarise suggestions and that name change will be made.

Often name changes are a bit annoying though if it's easy to change name now and agree.

Reply:
Ok took a wack at a name change in https://github.com/CDCgov/cfa-epinow2-pipeline/pull/79/commits/5f132b1a0a029eadb9079585eec4eedc7749d70e

Let's revisit in the next PR?